src/machine/machine-varlink.c

```diff
@@ -375,3 +378,195 @@ int vl_method_kill(sd_varlink *link, sd_json_variant *parameters, sd_varlink_met
+        if (r < 0)
+                return r;
+
+        r = varlink_verify_polkit_async(
+                        link,
+                        manager->bus,
+                        machine_open_polkit_action(p.mode, machine->class),
+                        machine_open_polkit_details(p.mode, machine->name, user, path, command_line),
+                        &manager->polkit_registry);
+        if (r <= 0)
+                return r;
+
+        ptmx_fd = machine_openpt(machine, O_RDWR|O_NOCTTY|O_CLOEXEC, &ptmx_name);
+        if (ERRNO_IS_NEG_NOT_SUPPORTED(ptmx_fd))
+                return sd_varlink_error(link, "io.systemd.Machine.NotSupported", NULL);
+        if (ptmx_fd < 0)
+                return log_debug_errno(ptmx_fd, "Failed to open pseudo terminal: %m");
+
+        switch (p.mode) {
+                case MACHINE_OPEN_MODE_TTY:
+                        /* noop */
+                        break;
+
+                case MACHINE_OPEN_MODE_LOGIN:
+                        r = machine_start_getty(machine, ptmx_name, /* error = */ NULL);
+                        if (r == -ENOENT)
+                                return sd_varlink_error(link, "io.systemd.Machine.NoIPC", NULL);
+                        if (ERRNO_IS_NEG_NOT_SUPPORTED(r))
+                                return sd_varlink_error(link, "io.systemd.Machine.NotSupported", NULL);
+                        if (r < 0)
+                                return log_debug_errno(r, "Failed to start getty for machine '%s': %m", machine->name);
+
+                        break;
+
+                case MACHINE_OPEN_MODE_SHELL: {
+                        assert(user && path && args); /* to avoid gcc complaining about possible uninitialized variables */
+                        r = machine_start_shell(machine, ptmx_fd, ptmx_name, user, path, args, p.env, /* error = */ NULL);
+                        if (r == -ENOENT)
+                                return sd_varlink_error(link, "io.systemd.Machine.NoIPC", NULL);
+                        if (ERRNO_IS_NEG_NOT_SUPPORTED(r))
+                                return sd_varlink_error(link, "io.systemd.Machine.NotSupported", NULL);
+                        if (r < 0)
+                                return log_debug_errno(r, "Failed to start shell for machine '%s': %m", machine->name);
+
+                        break;
+                }
+
+                default:
+                        assert_not_reached();
+        }
+
+        ptmx_fd_idx = sd_varlink_push_fd(link, ptmx_fd);
+        /* no need to handle -EPERM because we do sd_varlink_set_allow_fd_passing_output() above */
+        if (ptmx_fd_idx < 0)
+                return log_debug_errno(ptmx_fd_idx, "Failed to push file descriptor over varlink: %m");
+
+        TAKE_FD(ptmx_fd);
+
+        r = sd_json_buildo(
+                        &v,
+                        SD_JSON_BUILD_PAIR_UNSIGNED("ptyFileDescriptor", ptmx_fd_idx),
```

Comment:
doesn't matter, but strictly speaking ptmx_fd_idx is a signed entity, and you pass it here via SD_JSON_BUILD_PAIR_UNSIGNED(). Which is a bit weird. Not a real issue, because you verified above that this is not negative, but I still would make this SD_JSON_BUILD_PAIR_INTEGER() to avoid any raised eyebrows.